File: lists/arrayList/c/intArrayList.c
//
// intArrayList.c
// By Scott Hall
//
// Models the data structure for an Integer type ArrayList

#include <stdio.h>
#include <stdlib.h>
#include "intArrayList.h"

// An initializer (sort of a constructor) for integer array list
void IntArrayList(int size, intArrayList *arrayList) {

    arrayList->maxSize = size;
    arrayList->length  = 0;
    arrayList->aList   = (int *) calloc(size, sizeof(int));

}

// --------------------------------------------------------------------------

// Return the maximum allowed size of the ArrayList
int getMaxSize(intArrayList arrayList) {
    return arrayList.maxSize;
}

// --------------------------------------------------------------------------

// Return the current length of the array list
int getLength(intArrayList arrayList) {
    return arrayList.length;
}

// --------------------------------------------------------------------------

// Returns whether the list is empty or not
int isEmpty(intArrayList arrayList) {
    if (arrayList.length == 0) {
        return 1;
    } 
    return 0;
}

// --------------------------------------------------------------------------

// Returns whether the list is full
int isFull(intArrayList arrayList) {
    if (arrayList.length == arrayList.maxSize) {
        return 1;
    }
    return 0;
}

// --------------------------------------------------------------------------

// Clear the ArrayList
void clear(intArrayList arrayList) {
    arrayList.length = 0;
}

// --------------------------------------------------------------------------

// Check if a specified item is in the ArrayList.  Returns the index of the
// match or a -1 if the item is not in the list
int searchArrayList(int value, intArrayList arrayList) {
    int i;

    for (i=0; i<arrayList.length; i++) {
        if (value == arrayList.aList[i]) {
            return i;
        }
    }
    return -1;
}

// --------------------------------------------------------------------------

// Insert an item at the end of the list if it does not already exist
int insertAtEnd(int value, intArrayList *arrayList) {

    if (isFull(*arrayList) == 1) {
        printf("insertAtEnd Error: List is already full.  Cannot add new item\n");
        return 1;
    }    

    if (searchArrayList(value, *arrayList) != -1) {
        printf("insertAtEnd Error: List is already has %d.  Cannot add new item\n", value);
        return 1;
    }

    arrayList->aList[arrayList->length] = value;
    arrayList->length++;

    return 0;
}

// --------------------------------------------------------------------------

// Insert an item at a specified index.  Index must exist, item not already
// exist and the list is not full for item to be added.  All items are 
// shifted
int insertAtIndex(int index, int value, intArrayList *arrayList) {

    int i;

    if (isFull(*arrayList) == 1) {
        printf("insertAtEnd Error: List is already full.  Cannot add new item\n");
        return 1;
    }    

    if (searchArrayList(value, *arrayList) != -1) {
        printf("insertAtEnd Error: List is already has %d.  Cannot add new item\n", value);
        return 1;
    }

    if (index > arrayList->length || index < 0) {
        printf("insertAtEnd Error: Index is out of range\n");
        return 1;   
    }

    arrayList->length++;

    for (i=arrayList->length; i>=index; i--) {
        arrayList->aList[i+1] = arrayList->aList[i];
    }

    arrayList->aList[index] = value;

    return 0;

}

// --------------------------------------------------------------------------

// Check if an item at a given index is equal to value
int isItemAtEqual(int index, int value, intArrayList arrayList) {

    if (index > arrayList.length || index < 0) {
        printf("isItemAtEqual Error: Index is out of range\n");
        return -1;   
    }

    if (arrayList.aList[index] == value) {
        return 1;
    }

    return 0;

}

// --------------------------------------------------------------------------

// Remove item at speficied index
int removeAtIndex(int index, intArrayList *arrayList) {

    int i;

    if (index > arrayList->length || index < 0) {
        printf("insertAtEnd Error: Index is out of range\n");
        return 1;   
    }

    // Shift contents over
    for (i=index; i<arrayList->length; i++) {
        if (i+1 >= getMaxSize(*arrayList)) {
            break;
        }
        arrayList->aList[i] = arrayList->aList[i+1];
    }

    arrayList->length--;

    return 0;

}

// --------------------------------------------------------------------------

// Remove specified item from list (if it exists)
int removeItem(int value, intArrayList *arrayList) {

    int i;
    int index;

    // Get index of value in the list
    index = searchArrayList(value, *arrayList);

    if (index == -1) {
        printf("removeItem Error: List does not contain %d.  Cannot remove item\n", value);
        return 1;
    }

    // Shift contents over
    for (i=index; i<arrayList->length; i++) {
        arrayList->aList[i] = arrayList->aList[i+1];
    }

    arrayList->length--;

    return 0;

}

// --------------------------------------------------------------------------

// Remove specified item from list (if it exists)
int getItemAtIndex(int index, int *value, intArrayList arrayList) {

    if (index > arrayList.length || index < 0) {
        printf("getItemAtIndex Error: Index is out of range\n");
        return -1;   
    }

    *value = arrayList.aList[index];

    return 0;

}

// --------------------------------------------------------------------------

// Print the list contents
void printList(intArrayList arrayList) {

    int i;

    printf("Integer ArrayList:\n");

    for (i=0; i<arrayList.length; i++) {

        if (i>0 && i%10==10) {
            printf("\n");
        }

        printf("%10d, ", arrayList.aList[i]);

    }

    printf("\n");
}

// --------------------------------------------------------------------------

// Free the dynamically allocated ArrayList
void freeList(intArrayList *arrayList) {
    free(arrayList->aList);
}

// --------------------------------------------------------------------------

int main() {

    int i;
    int value;

    // Create and initialize an array list
    intArrayList arrayList1;
    intArrayList arrayList2;

    // Initialize the array lists
    IntArrayList(20, &arrayList1);
    IntArrayList(10, &arrayList2);

    for (i=0; i<10; i++) {
        insertAtEnd(i, &arrayList1);
        insertAtEnd(i, &arrayList2);
    }

    printf("ArrayList 1:\n");
    printList(arrayList1);
    printf("List length size %d and max size %d\n", getLength(arrayList1), getMaxSize(arrayList1));

    printf("\nLet's add 99 at index 2\n"); // Should work
    insertAtIndex(2,99, &arrayList1);
    printList(arrayList1);

    printf("\nLet's try to add 99 at index 2 again\n"); // Should fail, we already added 99
    insertAtIndex(2,99, &arrayList1);
    printList(arrayList1);

    printf("\nLet's check if index 2 has 99 in it\n"); 
    printf("Is 99 at index 2? %d\n", isItemAtEqual(2,99, arrayList1));

    printf("\nLet's remove the value 99\n");
    removeItem(99, &arrayList1);
    printList(arrayList1);

    printf("\nLet's fetch the item at index 7\n");
    printf("Did I find something? %d Better not get -1\n", getItemAtIndex(7, &value, arrayList1));
    printf("The value fetched is %d\n", value);


    printf("\n\nArrayList 2:\n");
    printList(arrayList2);
    printf("List length size %d and max size %d\n", getLength(arrayList2), getMaxSize(arrayList2));

    printf("\nLet's add 99 at index 2\n"); // Should fail, list is full
    insertAtIndex(2,99, &arrayList2);
    printList(arrayList2);

    printf("\nLet's remove the value at index 5\n");
    removeAtIndex(5, &arrayList2);
    printList(arrayList2);

    // Free the lists before quitting
    freeList(&arrayList1);
    freeList(&arrayList2);

    return 0;

}

File: lists/arrayList/go/src/arrayList/IntArrayList.go
package arrayList

import "fmt"

type intListStruct struct {
  maxSize int
  length  int
  aList   []int
}

// ----------------------------------------------------------------------------

// Initializes the array list and its structure components and returns the
// initialized struct
func IntArrayList(aSize int) intListStruct {

  var list intListStruct
  var indx int

  // Initialize the array list structure elements
  list.maxSize = aSize
  list.length  = 0

  indx = 0
  for indx < aSize {
      list.aList   = append(list.aList, 0)
      indx = indx+1
  }
  return list
}

// ----------------------------------------------------------------------------

// Returns if the list is is full
func IntArrayListIsFull(list intListStruct) bool {

  var result bool

  result = false

  if list.length == list.maxSize {
    result = true
  }
  return result

}

// ----------------------------------------------------------------------------

// Returns if the list is is empty
func IntArrayListIsEmpty(list intListStruct) bool {

  var result bool

  result = false

  if list.length == 0 {
    result = true
  }
  return result

}

// ----------------------------------------------------------------------------

// Returns the maximum size of the array list
func IntArrayListMaxSize(list intListStruct) int {
  return list.maxSize
}

// ----------------------------------------------------------------------------

// Returns the current length of the list
func IntArrayListLength(list intListStruct) int {
  return list.length
}

// ----------------------------------------------------------------------------

// Search for an item in the list and return the index if it exists
func SearchArrayList(value int, list intListStruct) int {

  var indx int

  for indx < list.length {
    if value == list.aList[indx] {
      return indx
    }
    indx++
  }

  return -1
}

// ----------------------------------------------------------------------------

// Adds an item at the end of the list.  The item cannot already exist and
// the list cannot be full.
func InsertAtEnd(value int, list *intListStruct) int {

  // Is the list full?
  if IntArrayListIsFull(*list) == true {
    fmt.Println("insertAtEnd Error: List is full cannot add new item")
    return 1
  }

  // Does the list already have the value
  if SearchArrayList(value, *list) > -1 {
    fmt.Println("insertAtEnd Error: List already has the value ", value)
    return 1
  }

  list.aList[list.length] = value
  list.length++

  return 0
}

// ----------------------------------------------------------------------------

// Adds an item into the list at the specified index.  If the list is full,
// or the value is already in the list or the index given is invalid then
// nothing is added
func InsertAtIndex(index int, value int, list *intListStruct) int {

  var i int

  // Is the list full?
  if IntArrayListIsFull(*list) == true {
    fmt.Println("insertAtEnd Error: List is full cannot add new item")
    return 1
  }

  // Is the index valid?
  if (index < 0 || index > list.length) {
    fmt.Println("insertAtEnd Error: Index given is outside the range of list")
    return 1
  }

  // Does the list already have the value
  if SearchArrayList(value, *list) > -1 {
    fmt.Println("insertAtEnd Error: List already has the value ", value)
    return 1
  }

  // Shift the contents over by one after index
  i = list.length
  for i > index {
    list.aList[i] = list.aList[i-1]
    i--
  }

  list.aList[index] = value
  list.length++

  return 0
}

// ----------------------------------------------------------------------------

// Removes the item at specified index.  The index must be valid
func RemoveAtIndex(index int, list *intListStruct) int {

  var i int

  // Is the index valid?
  if (index < 0 || index > list.length) {
    fmt.Println("insertAtEnd Error: Index given is outside the range of list")
    return 1
  }

  // Shift the contents over by one after index
  i = index
  for i < list.length-1 {
    list.aList[i] = list.aList[i+1]
    i++
  }

  list.length--

  return 0
}

// ----------------------------------------------------------------------------

// Searchs for a value in the list and removes it.  If the item does not exist
// then nothing is done.
func Remove(value int, list *intListStruct) int {

  var i     int
  var index int

  // Get the index of the value in the list.  If -1 is returned then the
  // value does not exist in the list
  index = SearchArrayList(value, *list)

  // Is the index valid?
  if (index == -1) {
    fmt.Println("Remove Error: Value ", value, " does not exist in list.  Cannot remove")
    return 1
  }

  // Shift the contents over by one after index
  i = index
  for i < list.length-1 {
    list.aList[i] = list.aList[i+1]
    i++
  }

  list.length--

  return 0
}

// ----------------------------------------------------------------------------

// Searchs for a value in the list and removes it.  If the item does not exist
// then nothing is done.  Returns a map which identifies if value was found
// and the value itself.  If not found then zero's are returned
func GetItemAtIndex(index int, list intListStruct) map[string]int {

  result := make(map[string]int)

  // Is the index valid?
  if (index < 0 || index > list.length) {
    fmt.Println("GetItemAtIndex Error: Index is not in valid range")
    result["valid"] = 0
    result["value"] = 0
    return result
  }

  result["valid"] = 1
  result["value"] = list.aList[index]

  return result
}

// ----------------------------------------------------------------------------

// Check if an item is equal at an index
func IsItemAtEqual(index int, value int, list intListStruct) bool {

  // Is the index valid?
  if (index < 0 || index > list.length) {
    fmt.Println("IsItemAtEqual Error: Index is not in valid range")
    return false
  }

  if list.aList[index] == value {
    return true
  }

  return false
}

// ----------------------------------------------------------------------------

// Print the contents of the array list
func PrintIntArrayList(list intListStruct) {

  var indx int

  fmt.Println("INTEGER ARRAY LIST: ")

  for indx < list.length {
    if indx % 10 == 0 && indx != 0 {
      fmt.Println("")
    }
    fmt.Print(list.aList[indx], ", ")
    indx = indx + 1
  }
  fmt.Println("")

}
